ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs

```diff
@@ -137,7 +137,31 @@ data Cmd blk it rdr
   | RunBgTasks
   deriving (Generic, Show, Functor, Foldable, Traversable)
 
--- TODO knownInvalidBlocks, see #625
+-- We don't test 'getIsInvalidBlock' because the simple chain selection in the
+-- model and the incremental chain selection in the real implementation differ
+-- non-trivially.
+--
+-- In the real chain selection, if a block is invalid, no chains containing
+-- that block will be validated again. So if a successor of the block is added
+-- afterwards, it will never be validated, as any chain it would be part of
+-- would also contain the block we know to be invalid. So if the successor is
+-- also invalid, we would never discover and record it (as there is no need
+-- from the point of chain selection, as we'll never use it in a candidate
+-- chain anyway). In the model implementation of chain selection, all possible
+-- chains are (re)validated each time, including previously invalid chains, so
+-- new invalid blocks that are successors of known invalid blocks /are/ being
+-- validated and recorded as invalid blocks.
+--
+-- Further complicating this is the fact that the recorded invalid blocks are
+-- also garbage-collected. We can work around this, just like for 'getBlock'.
+--
+-- While it is certainly possible to overcome the issues described above,
+-- e.g., we could change the model to closer match the real implementation
+-- (but at the cost of more complexity), it is not worth the effort. The whole
+-- point of recording invalid blocks is to avoid constructing candidates
+-- containing known invalid blocks and needlessly validating them, which is
+-- something we are testing in 'prop_trace', see
+-- 'invalidBlockNeverValidatedAgain'.
```

Comment:
Nice explanation.

Reply:
> How frequently is this called, once per block or slot? The Map filter is linear. There's a trade-off for how frequently we scan vs how much excess memory we use.

Once per block (using that block's slot).

Ideally, we'd have:
* *O(log n)* lookup of `HeaderHash blk`
* *O(log n)* filtering (`Map.split`) based on `SlotNo`

Unfortunately, we can't have both, so the latter is indeed *O(n)*.

We continue thinking about this in #730.

Reply:
Ok, happy to defer that one to #730.